jax/experimental/sparse/api.py

```diff
@@ -107,6 +108,95 @@ def _todense_batching_rule(batched_args, batch_dims, *, tree):
 mlir.register_lowering(todense_p, mlir.lower_fun(
     _todense_impl, multiple_results=False))
 
+#----------------------------------------------------------------------
+# map_specified_elements – function to map a scalar function to specified elements
+map_specified_elements_p = core.Primitive('map_specified_elements')
+map_specified_elements_p.multiple_results = True
+
+def map_specified_elements(fun: Callable[[ArrayLike], ArrayLike], arg: Union[ArrayLike, JAXSparse]) -> Union[ArrayLike, JAXSparse]:
+  """Map a given function across the specified elements of an array
+
+  If the input is a dense array, ``fun`` will be applied to every array element.
+  If the input is a sparse array, ``fun`` will be applied only to the data buffer.
```

Comment:
> [...] it must also accept dense inputs so that it can be used within a function that is transformed by sparsify

I understand that we want to express this operation in "dense mode", but do we actually want it to execute and have semantics other than "error" in that case? As the comment here notes, the meaning is (possibly subtly) different for dense vs sparse operands.

Reply:
My thought on this was that for dense inputs, every element is specified! So this lets you write one function that applies to both sparse and dense inputs (either with direct sparse inputs or implicitly via `sparsify`)

Reply:
In this case, `sparsify(f)` may no longer be a semantic identity, for example if:
```python
f = partial(sparse.map_specified_elements(jnp.cos)
```
Right? It still won't be a semantic identity if it errs vs. runs too, but I'm offering that that might be better than silently-not-being-the-identity.

Reply:
Note you can still `sparsify` a function that calls `map_specified_elements` and have `map_specified_elements` error on execution. It's a matter of a primitive having a valid sparsification rule while, say, its impl/lowering rule is an error.